components/compliance-service/reporting/relaxting/suggestions.go

```diff
@@ -502,6 +502,170 @@ func (backend ES2Backend) getControlSuggestions(client *elastic.Client, typePara
+		Query(boolQuery).
+		FetchSource(false).
+		Size(size * 50) // Multiplying size to ensure that same profile with multiple versions is not limiting our suggestions to a lower number
+		// ^ Because we can't sort by max_score of the inner hits: https://discuss.elastic.co/t/nested-objects-hits-inner-hits-and-sorting/32565
+
+	source, err := searchSource.Source()
+	if err != nil {
+		return nil, errors.Wrap(err, "getControlTagsSuggestions unable to get Source")
+	}
+
+	LogQueryPartMin(esIndex, source, "getControlTagsSuggestions query")
+
+	searchResult, err := client.Search().
+		SearchSource(searchSource).
+		Index(esIndex).
+		FilterPath(
+			"took",
+			"hits.total",
+			"hits.hits._id",
+			"hits.hits._score",
+			"hits.hits.inner_hits").
+		Do(context.Background())
+
+	if err != nil {
+		logrus.Error("getControlTagsSuggestions search failed")
+		return nil, err
+	}
+
+	logrus.Debugf("Search query took %d milliseconds\n", searchResult.TookInMillis)
+
+	type ControlSourceTags struct {
+		ID         string                            `json:"id"`
+		StringTags []ESInSpecReportControlStringTags `json:"string_tags"`
+	}
+	type ControlTagsScore struct {
+		Score      float32
+		StringTags []ESInSpecReportControlStringTags
+	}
+
+	// Using this array to aggregate the unsorted suggestions from the response
+	foundControlTags := make([]ControlTagsScore, 0)
+	if searchResult != nil {
+		for _, hit := range searchResult.Hits.Hits {
+			if hit == nil {
+				continue
+			}
+			for _, inner_hit := range hit.InnerHits {
+				if inner_hit == nil {
+					continue
+				}
+				for _, hit2 := range inner_hit.Hits.Hits {
+					if hit2.Source == nil {
+						continue
+					}
+					var c ControlSourceTags
+					err := json.Unmarshal(*hit2.Source, &c)
+					if err == nil && c.ID != "" {
+						foundControlTags = append(foundControlTags, ControlTagsScore{
+							float32(*hit2.Score),
+							c.StringTags,
+						})
+					}
+				}
+			}
+		}
+	}
+	sort.Slice(foundControlTags, func(i, j int) bool {
+		return foundControlTags[i].Score > foundControlTags[j].Score
+	})
+
+	// Map the found tag suggestions to remove duplications
+	scoredTagSuggs := make(map[string]*float32)
+	for _, item := range foundControlTags {
+		matches := make([]string, 0)
+		// If the suggestion is for control tag key, we need to find the one that
+		// was the closest match from all the tag keys the control might have
+		if typeParam == "control_tag_key" {
+			for _, controlTag := range item.StringTags {
+				matches = append(matches, controlTag.Key)
+			}
+		}
+
+		// If the suggestion is for control tag values, we need to find the tag that matches the
+		// tag key and then find the closest match from all the values of that tag
+		if typeParam == "control_tag_value" {
+			for _, controlTag := range item.StringTags {
+				if controlTag.Key == controlTagFilterKey {
+					matches = append(matches, controlTag.Values...)
+				}
+			}
+		}
+
+		// Find the best engram match from the array of texts
```

Comment:
I think there is some subtly here that might be good to put into a comment.  Namely, `firstBestArrayMatch` either returns an array containing only the best match OR (in the case of len(text) < 2) it just returns us the original array.  

Since item.Score is invariant over the duration of 629-634 I _think_ that loop is equivalent to just looking at the first item if it exists.

My guess is if we were to look at the assembly the compiler has recognized all of this, but...I'm thinking there might be a way to restructure the code here a _bit_ to make some of this implicit behavior a bit more explicit.